pkg/instrument/dqlite.go

```diff
@@ -0,0 +1,125 @@
+package instrument
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"sync"
+	"sync/atomic"
+)
+
+type Listener struct {
+	Network, Address        string
+	AcceptedChan            chan net.Conn
+	netListener             net.Listener
+	bytesWritten, bytesRead atomic.Int64
+	wg                      sync.WaitGroup
+	err                     error
+}
+
+type ListenerMetrics struct {
+	BytesWritten int64
+	BytesRead    int64
+}
+
+func NewListener(network, address string) *Listener {
+	return &Listener{
+		Network:      network,
+		Address:      address,
+		AcceptedChan: make(chan net.Conn),
+		netListener:  nil,
+	}
+}
+
+func (l *Listener) Listen(ctx context.Context) error {
+	if l.netListener != nil {
+		return fmt.Errorf("listener already running")
+	}
+	var lc net.ListenConfig
+	netListener, err := lc.Listen(ctx, l.Network, l.Address)
```

Comment:
A good convention to keep to is to use `var v T` declarations when `v` is later assigned to before being read. I think it would cleaner to explicitly assign this to its zero value to indicate that it will be used in that state

Also the name `lc` is quite short, could this be made more descriptive, perhaps as `netListenerConfig`? I'm on the fence about this one

Reply:
TBH I copied this snippet [from the Go runtime](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/dial.go;l=806), so I don't really have any opinion on this.

I'll go with both suggestions.